### PR TITLE
Refactor provider shutdown and pending flows

### LINKED ISSUE #226: flaky test: HTTPClientTests.testRaceNewRequestsVsShutdown
5.1 CI

```
Test Case 'HTTPClientTests.testRaceNewRequestsVsShutdown' started at 2020-05-21 11:37:53.333
/code/Tests/AsyncHTTPClientTests/HTTPClientTests.swift:1139: error: HTTPClientTests.testRaceNewRequestsVsShutdown : failed - Unexpected error: ProviderClosedError()
Assertion failed: file /code/Sources/AsyncHTTPClient/ConnectionsState.swift, line 92
Current stack trace:
0    libswiftCore.so                    0x00007ffaa5df032b <unavailable> + 3953451
1    libswiftCore.so                    0x00007ffaa5df03a0 swift_reportError + 50
2    libswiftCore.so                    0x00007ffaa5e61090 _swift_stdlib_reportFatalErrorInFile + 115
3    libswiftCore.so                    0x00007ffaa5d854fe <unavailable> + 3515646
4    libswiftCore.so                    0x00007ffaa5d85677 <unavailable> + 3516023
5    libswiftCore.so                    0x00007ffaa5b7addd <unavailable> + 1375709
6    libswiftCore.so                    0x00007ffaa5d5c458 <unavailable> + 3347544
7    libswiftCore.so                    0x00007ffaa5b7a179 <unavailable> + 1372537
8    async-http-clientPackageTests.xctest 0x000055d0f56dbaeb <unavailable> + 1346283
9    async-http-clientPackageTests.xctest 0x000055d0f56c8bd1 <unavailable> + 1268689
10   async-http-clientPackageTests.xctest 0x000055d0f56c8d04 <unavailable> + 1268996
11   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
12   async-http-clientPackageTests.xctest 0x000055d0f56c66a3 <unavailable> + 1259171
13   async-http-clientPackageTests.xctest 0x000055d0f56c6704 <unavailable> + 1259268
14   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
15   async-http-clientPackageTests.xctest 0x000055d0f56d5a24 <unavailable> + 1321508
16   async-http-clientPackageTests.xctest 0x000055d0f56d5dd3 <unavailable> + 1322451
17   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
18   async-http-clientPackageTests.xctest 0x000055d0f56d5e92 <unavailable> + 1322642
19   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
20   async-http-clientPackageTests.xctest 0x000055d0f5d910e2 <unavailable> + 8380642
21   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
22   async-http-clientPackageTests.xctest 0x000055d0f5d7f04f <unavailable> + 8306767
23   async-http-clientPackageTests.xctest 0x000055d0f5d825cb <unavailable> + 8320459
24   async-http-clientPackageTests.xctest 0x000055d0f5d823a5 <unavailable> + 8319909
25   async-http-clientPackageTests.xctest 0x000055d0f56f95c5 <unavailable> + 1467845
26   async-http-clientPackageTests.xctest 0x000055d0f5706cc8 <unavailable> + 1522888
27   async-http-clientPackageTests.xctest 0x000055d0f56cae18 <unavailable> + 1277464
28   async-http-clientPackageTests.xctest 0x000055d0f5706d61 <unavailable> + 1523041
29   async-http-clientPackageTests.xctest 0x000055d0f5d9b621 <unavailable> + 8422945
30   async-http-clientPackageTests.xctest 0x000055d0f5d9cd66 <unavailable> + 8428902
31   async-http-clientPackageTests.xctest 0x000055d0f5d8cf64 <unavailable> + 8363876
32   async-http-clientPackageTests.xctest 0x000055d0f5d91140 <unavailable> + 8380736
33   async-http-clientPackageTests.xctest 0x000055d0f5d878af <unavailable> + 8341679
34   async-http-clientPackageTests.xctest 0x000055d0f5d8bb23 <unavailable> + 8358691
35   async-http-clientPackageTests.xctest 0x000055d0f5d907f0 <unavailable> + 8378352
36   async-http-clientPackageTests.xctest 0x000055d0f5e419e8 <unavailable> + 9103848
37   async-http-clientPackageTests.xctest 0x000055d0f5e4961e <unavailable> + 9135646
38   async-http-clientPackageTests.xctest 0x000055d0f5e41a38 <unavailable> + 9103928
39   async-http-clientPackageTests.xctest 0x000055d0f5e46f9c <unavailable> + 9125788
40   async-http-clientPackageTests.xctest 0x000055d0f5e41ab5 <unavailable> + 9104053
41   async-http-clientPackageTests.xctest 0x000055d0f5e46e8c <unavailable> + 9125516
42   async-http-clientPackageTests.xctest 0x000055d0f56f0340 <unavailable> + 1430336
43   async-http-clientPackageTests.xctest 0x000055d0f5e46ef8 <unavailable> + 9125624
44   async-http-clientPackageTests.xctest 0x000055d0f5e38059 <unavailable> + 9064537
45   async-http-clientPackageTests.xctest 0x000055d0f5e3f5c4 <unavailable> + 9094596
46   async-http-clientPackageTests.xctest 0x000055d0f5d71780 <unavailable> + 8251264
47   async-http-clientPackageTests.xctest 0x000055d0f5d7257d <unavailable> + 8254845
48   async-http-clientPackageTests.xctest 0x000055d0f5d7e1d1 <unavailable> + 8303057
49   async-http-clientPackageTests.xctest 0x000055d0f5d72f5c <unavailable> + 8257372
50   async-http-clientPackageTests.xctest 0x000055d0f5eb6909 <unavailable> + 9582857
51   async-http-clientPackageTests.xctest 0x000055d0f5ebc34b <unavailable> + 9605963
52   async-http-clientPackageTests.xctest 0x000055d0f5ebc6b1 <unavailable> + 9606833
53   async-http-clientPackageTests.xctest 0x000055d0f5650003 <unavailable> + 774147
54   libpthread.so.0                    0x00007ffaa58156ba <unavailable> + 30394
55   libc.so.6                          0x00007ffaa39ff3b0 clone + 109
Exited with signal code 4
Build step 'Execute shell' marked build as failure
```

@artemredkin could this be a bug in the new pool implementation?

### PR DESCRIPTION
**Pull request by artemredkin**
Fixes some edge cases in shutdown and pending states.

Motivation:
There are two issues this PR addresses:
1. Flaky test, where we are trying to close the client when connection was able to `enqueue` and get an unexpected error
2. The fact that provider is created with `+1` pending state, while it might not be an error, it makes testing difficult, also might lead to races in shutdown.

Modifications:
1. `HTTP1ConnectionsProvider` is no longer created with `+1` pending
2. If we receive `close` while request is enqueued, it should be failed with a more expected `.alreadyShutdown` error, since its the only way to get into that state, so test was actually catching the right thing.

Result:
Closes #226 
Closes #323

### CODE REVIEW

File: Sources/AsyncHTTPClient/ConnectionPool.swift [hunk 1]
```diff
@@ -80,12 +80,12 @@ final class ConnectionPool {
             if let existing = self.providers[key], existing.enqueue() {
                 return existing
             } else {
-                // Connection provider will be created with `pending = 1`
                 let provider = HTTP1ConnectionProvider(key: key,
                                                        eventLoop: taskEventLoop,
                                                        configuration: self.configuration,
                                                        pool: self,
                                                        backgroundActivityLogger: self.backgroundActivityLogger)
+                _ = provider.enqueue()
```
Comment (weissi) [inline Sources/AsyncHTTPClient/ConnectionPool.swift:88]: yippie! Could we assert here that the return value is the expected one?

File: Sources/AsyncHTTPClient/ConnectionsState.swift [hunk 2]
```diff
@@ -64,7 +64,7 @@ extension HTTP1ConnectionProvider {
         private var openedConnectionsCount: Int = 0
 
         /// Number of enqueued requests, used to track if it is safe to delete the provider.
-        private var pending: Int = 1
+        private var pending: Int = 0
```
Comment (weissi) [inline Sources/AsyncHTTPClient/ConnectionsState.swift:67]: nice! I left a bunch of stuff like the following comments in the code

```
        /* BEGIN workaround for #232, this whole block is to be replaced by the commented out line below */
        // not closing the provider here (https://github.com/swift-server/async-http-client/issues/232)
        var state = self.http1ConnectionProvider.state.testsOnly_getInternalState()
        if state.pending == 1, state.waiters.isEmpty, state.leasedConnections.isEmpty, state.openedConnectionsCount == 0 {
            state.pending = 0
            self.http1ConnectionProvider.state.testsOnly_setInternalState(state)
        }
        self.http1ConnectionProvider.closePromise.succeed(())
        /* END workaround for #232 */
```

and

```
            // cleanup
            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
```

I believe we can also delete all this code then.

File: Sources/AsyncHTTPClient/ConnectionPool.swift (hunk 1 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/ConnectionPool.swift]: great idea, done!

File: Sources/AsyncHTTPClient/ConnectionsState.swift (hunk 2 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/ConnectionsState.swift:67]: I was able to delete the workaround, some clean up code, and I was able to replace some cleanup with API calls instead of direct state manipulation. Currently there are 3-4 tests that manipulate state directly (they all need the state to be near the limit), do you want me to fix it here or when I get to #232?

Comment (weissi) [reply Sources/AsyncHTTPClient/ConnectionsState.swift:67]: no, let's fix them in a followup. It's super awesome you could remove the workaround

File: Sources/AsyncHTTPClient/ConnectionsState.swift [hunk 3]
```diff
@@ -148,7 +148,7 @@ extension HTTP1ConnectionProvider {
                     return .none
                 }
             case .closed:
-                return .fail(waiter, ProviderClosedError())
+                return .fail(waiter, HTTPClientError.alreadyShutdown)
```
Comment (weissi) [inline Sources/AsyncHTTPClient/ConnectionsState.swift:151]: @artemredkin from what I can tell, no test noticed that this error changed which probably means no test is getting this error ;). I think we should probably add a test case that tests this condition.

Comment (weissi) [reply Sources/AsyncHTTPClient/ConnectionsState.swift:151]: filed #244

Comment (artemredkin) [reply Sources/AsyncHTTPClient/ConnectionsState.swift:151]: Specific test is a good idea, yes, though this error was picked up by a test, just not reliably :)
